feat(client-debugger): Add system for generating visual summary trees for DDS_s

Josmithr opened the pull request:
This PR adds a system to the debugger which can generate visual summary trees for DDS_s provided to the debugger (as well as nested DDS_s).

### Visual Tree Format

Here, I have taken the design approach of having the debugger generate visual descriptions of DDS contents. Consumers of these visual trees can then process the visual data as they see fit, but the assumption is that they will be rendered in some sort of tree/graph-like representation.

See `VisualTree.ts` for the specific types.

Consumers can pattern match on the `NodeKind` associated with each node.

Data can be requested for a `FluidHandleNode` via the `GET_DATA_VISUALIZATION` message.
This will yield a `DATA_VISUALIZATION` message containing a `FluidObjectNode` tree to be rendered.

### Messaging Flow

The following sample can be found in `Overview.md`. It offers a high-level summary of the message-passing flow for "rendering" visual trees describing DDS contents.

```mermaid
sequenceDiagram
participant Application
participant Debugger
participant Consumer

Application->>Debugger: Initialize debugger
Consumer-->>Debugger: "GET_ROOT_DATA_VISUALIZATIONS"
Debugger->>Consumer: "ROOT_DATA_VISUALIZATIONS" ([handle1, handle2, ..., handleN])
loop renderTrees
	Consumer->>Consumer: renderTree(handle.id)
end
Consumer-->>Debugger: "GET_DATA_VISUALIZATION" (id)
Debugger->>Consumer: "DATA_VISUALIZATION" (visualization)
```

Initial rendering (and subsequent automatic updates) can be initiated by the consumer via the `GET_ROOT_DATA_VISUALIZATIONS` message. This will generate a list of `FluidHandleNode`s for any `rootData` provided to the debugger at initialization time.

While recursively rendering visualization trees, any time a `FluidHandleNode` is encountered, its corresponding data can be requested via `GET_DATA_VISUALIZATION`.

Once a DDS's data has been requested this way for the first time, subsequent changes will automatically result in `DATA_VISUALIZATION` messages being posted, so consumers will not need to poll for updates or anything like that. So long as they subscribe to the `DATA_VISUALIZATION` message, they will receive updates automatically.

### Notes for Reviewers

Primary points of note for the review:

- `VisualTree.ts`: Contains a grammar for describing visual summary trees for DDS data. Interested in thoughts on type names here as well as the underlying data format.
- `DataVisualization.ts`": Contains a lazy / incremental system for generating visual summary trees. The entry-point (used initialized by the debugger) is the `DataVisualizerGraph`. Generates `VisualNode`s on demand on a per-DDS granularity. 

### Follow-Up Work

The current implementation has some limitations that should be fixed before MVP.

The most important issue currently is that, once a visualization has been requested for a DDS, we never clean that visualizer node up. So even if the associated DDS becomes unreachable by walking `rootData`, we currently will never dispose of the corresponding node, and will continue to broadcast updates. For applications that create many dynamic DDSs, this may result in non-trivial memory leakage. We will need to add some dependency tracking to ensure we can clean up resources as needed. The plan is to do this in a subsequent PR.

We also will likely want to broadcast visual tree diffs, rather than complete trees, when posting updates. But we can address this as performance demands.

These follow-up considerations are also noted in `Overview.md`.

Josmithr committed:
docs: Add comment

Josmithr committed:
[WIP]

Josmithr committed:
Merge branch 'main' into client-debugger/schema-based-data-prototype

Josmithr committed:
style: Prettier

Josmithr committed:
feat: Add visual tree format

Josmithr committed:
refactor: Update prototype to use new tree format

Josmithr committed:
WIP

Josmithr committed:
WIP

Josmithr committed:
Merge branch 'main' into client-debugger/schema-based-data-prototype

Josmithr committed:
build: Add dep

Josmithr committed:
WIP

Josmithr committed:
WIP

Josmithr committed:
WIP

Josmithr committed:
fix: Make vis node disposable

Josmithr committed:
refactor: Renam,e class

Josmithr committed:
WIP

Josmithr committed:
fix: Return types

Josmithr committed:
feat: Root rendering support

Josmithr committed:
refactor: Rename module

Josmithr committed:
refactor: Rename directory

Josmithr committed:
build: Add test dependency

Josmithr committed:
refactor: Move helper function

Josmithr committed:
feat: Add default visualizers const

Josmithr committed:
feat: Add directory exports

Josmithr committed:
test: Add simple unit test

Josmithr committed:
test: Rename test

Josmithr committed:
refactor: Simplify DDS initialization

Josmithr committed:
feat: Add default visualizer for SharedCell

Josmithr committed:
feat: Add default visualizer for SharedString

Josmithr committed:
improvement: Add typeMetadata for object nodes

Josmithr committed:
test: Add multiple root DDS_s test

Josmithr committed:
refactor: Rename property

Josmithr committed:
refactor: Rename property

Josmithr committed:
docs: Add example comment

Josmithr committed:
docs: Add example comment

Josmithr committed:
docs: Punctuation

Josmithr committed:
docs: Update TODO

Josmithr committed:
feat: Export visual tree interfaces

Josmithr committed:
Merge branch 'main' into client-debugger/schema-based-data-prototype

Josmithr committed:
feat: Add data message types

Josmithr committed:
feat: Wire up message handling, add docs, and rename message types

Josmithr committed:
remove: Obsolete TODO

Josmithr committed:
docs: Add TODO

Josmithr committed:
docs: Add initial overview doc for data visualization

Josmithr committed:
docs: Simplify diagram a bit

Josmithr committed:
Merge branch 'main' into client-debugger/schema-based-data-prototype

Josmithr committed:
docs: More notes

Josmithr committed:
docs: Update TODO format

Josmithr committed:
docs: Add TODOs

Josmithr committed:
style: Prettier

Josmithr committed:
docs: Add comment

Josmithr committed:
refactor: Rename interface and add comment

Josmithr committed:
docs: Add missing `@inheritDoc` comments

Josmithr committed:
style: Prettier

Josmithr committed:
docs: Add comment

Josmithr committed:
refactor: Rename method and add documentation

Josmithr committed:
docs: Add method comment

Josmithr committed:
docs: More method comments

Josmithr committed:
docs: Add comments

Josmithr committed:
docs: Update TODOs

Josmithr committed:
docs: Update comments

Josmithr committed:
refactor: Rename property and add docs

Josmithr committed:
refactor: Simplify graph management

Josmithr committed:
docs: Update API report

Josmithr committed:
refactor: Rename constructor parameter

Josmithr committed:
docs: Add link

Josmithr committed:
docs: Add comment

Josmithr committed:
docs: Quote message type

Josmithr committed:
refactor: Update graph class to emit update events, and wire up to client debugger

Josmithr committed:
Merge branch 'main' into client-debugger/schema-based-data-prototype

Josmithr committed:
style: Sort package.json

Josmithr committed:
refactor: Rename class

Josmithr committed:
Merge branch 'main' into client-debugger/schema-based-data-prototype

Josmithr commented:
Note that `GET_ROOT_DATA_VISUALIZATIONS` returns a list of handle nodes, rather than full trees describing the root DDSs. This isn't strictly necessary, but it does have a couple of advantages:
1. We don't waste any time "rendering" child contents until explicitly requested (via `GET_DATA_VISUALIZATION`)
2. Enforces a pattern when consumers always start with handles, and have to explicitly make requests for the corresponding data (i.e. it makes the initial part of the flow less "special").

Interested in thoughts on this as well 🙂

Josmithr committed:
refactor: Rename tree interfaces

Josmithr committed:
refactor: Allow non-string "value" data and rename test module

Josmithr committed:
test: Add visualizer-level test

Josmithr committed:
test: Add SharedCell visualization test

Josmithr committed:
tools: Fix test:coverage script

Josmithr committed:
style: Prettier

Josmithr committed:
test: Add SharedString visualization test

Josmithr committed:
test: Add SharedMap visualization test

Josmithr committed:
test: Add visualizer test for unknown shared objects

Josmithr committed:
improvement: Include shared object "type" in generated metadata for unrecognized objects

Josmithr committed:
Merge branch 'main' into client-debugger/schema-based-data-prototype

Josmithr committed:
Merge branch 'main' into client-debugger/schema-based-data-prototype

Josmithr committed:
Merge branch 'main' into client-debugger/schema-based-data-prototype

Josmithr committed:
Merge branch 'main' into client-debugger/schema-based-data-prototype

Josmithr committed:
Merge branch 'main' into client-debugger/schema-based-data-prototype

Josmithr committed:
Merge branch 'main' into client-debugger/schema-based-data-prototype

Josmithr committed:
improvement: Add special node denoting an unrecognized Fluid object kind

Josmithr committed:
refactor: Add explicit return types

Josmithr committed:
improvement: Better type safety

Josmithr committed:
refactor: Expand metadata and value options

Josmithr committed:
feat: Add export

Josmithr committed:
fix: Add missing release tag

Josmithr committed:
improvement: Include `size` metadata in default `SharedMap` visualization

Josmithr committed:
docs: Update comment

Josmithr committed:
refactor: Explicit exports

Josmithr committed:
refactor: Type renames and helper type unions

Josmithr committed:
docs: Update API report

Josmithr committed:
Merge branch 'main' into client-debugger/schema-based-data-prototype

chentong7 commented on packages/tools/client-debugger/client-debugger/src/data-visualization/DataVisualization.ts:
```diff
@@ -0,0 +1,444 @@
+	 * If no such object exists in the graph, returns `undefined`.
+	 */
+	public async render(fluidObjectId: FluidObjectId): Promise<FluidObjectNode | undefined> {
+		if (!this.visualizerNodes.has(fluidObjectId)) {
```
```suggestion
		const visualizerNode = this.visualizerNodes.get(fluidObjectId);

    if (!visualizerNode) {
        // If there's no registered visualizer node, it might indicate a stale data request or a bug.
        // Return 'undefined' to indicate that the requested Fluid object could not be rendered.
```

Josmithr replied:
Good call. This method used to be a bit more complex, but ended up being simpler than I thought it needed to be. This can be a lot cleaner.

Josmithr replied:
Cleaned up in [6ed50a4](https://github.com/microsoft/FluidFramework/pull/14770/commits/6ed50a4b0d3e302b6d545630564e52c48ef35d42)

chentong7 commented on packages/tools/client-debugger/client-debugger/src/data-visualization/DataVisualization.ts:
```diff
@@ -0,0 +1,444 @@
+
+		// Checked above.
+		// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+		const visualizerNode = this.visualizerNodes.get(fluidObjectId)!;
```
```suggestion
		return visualizerNode.render();
```

chentong7 replied:
It might be better to make request to `visualizerNodes` only once

Josmithr replied:
Cleaned up in [6ed50a4](https://github.com/microsoft/FluidFramework/pull/14770/commits/6ed50a4b0d3e302b6d545630564e52c48ef35d42)

Josmithr committed:
refactor: Simplify render method

chentong7 reviewed (approved):
Approved with minor concern: During testing of different data types, casting some data objects to ISharedObject may fail, potentially resulting in errors and causing the page to break.

Josmithr commented on packages/tools/client-debugger/client-debugger/src/data-visualization/Overview.md:
```diff
@@ -0,0 +1,132 @@
+For the purpose of demonstrating the intended usage flow of the system, we will be looking at it from the context of the debugger, rather than viewing this system in isolation.
+
+When initializing the debugger, the initializing consumer may optionally provide a root DDS(s) they wish to have visualized by the tooling.
+If they do not specify this, then the tooling will not generate any visuals.
```
Maybe "visual descriptors"?

Josmithr commented on packages/tools/client-debugger/client-debugger/src/data-visualization/Overview.md:
```diff
@@ -0,0 +1,132 @@
+
+## Laziness
+
+The system has designed to be lazy.
```
🤦‍♂️

Josmithr commented on packages/tools/client-debugger/client-debugger/src/data-visualization/DefaultVisualizers.ts:
```diff
@@ -0,0 +1,136 @@
+		fluidObjectId: sharedCell.id,
+		label,
+		children: [renderedData],
+		typeMetadata: "SharedCell",
```
Possibly; I'll take a look.

Josmithr commented on pnpm-lock.yaml:
I'll double check.

Josmithr replied:
Yeah, apparently I updated pnpm at some point. Fixed now :)

Josmithr commented on packages/tools/client-debugger/client-debugger/src/data-visualization/DataVisualization.ts:
```diff
@@ -0,0 +1,438 @@
+	};
+
+	public constructor(
+		rootData: Record<string, IFluidLoadable>,
+		visualizers: SharedObjectVisualizers,
```
I'll admit I'm personally not a fan of the inline property syntax. But I like consistency. Will update 🙂

Josmithr commented on packages/tools/client-debugger/client-debugger/src/FluidClientDebugger.ts:
```diff
@@ -387,6 +456,11 @@ export class FluidClientDebugger
 		globalThis.removeEventListener?.("message", this.windowMessageHandler);
 
+		// Dispose of data visualization graph
+		// Note: no need to unbind associated event handler, since we're disposing the object altogether.
```
Ah, that's fair. I think you're right. Will update :)

Josmithr committed:
build: Remove unneeded dependency

Josmithr committed:
build: Remove unneeded dep

alexvy86 commented on packages/tools/client-debugger/client-debugger/src/data-visualization/DataVisualization.ts:
```diff
@@ -0,0 +1,438 @@
+	 * Adds a visualizer node to the collection for the specified
+	 * {@link @fluidframework/shared-object-base#ISharedObject} if one does not already exist.
+	 */
+	private registerVisualizerForSharedObject(
+		sharedObject: ISharedObject,
+		label: string,
+	): FluidObjectId {
+		if (!this.visualizerNodes.has(sharedObject.id)) {
```
... yeah ... that was the plan all along... 🙄 😝

Josmithr replied:
This got me thinking that "label" really doesn't make sense as a concept in our visual tree nodes. "Labels" are really a parent node concept (map key, property name, etc.). I restructured the tree grammar to omit "label" altogether, which also simplified the code a fair amount.

[eadae94](https://github.com/microsoft/FluidFramework/pull/14770/commits/eadae9422e0e7228b08b98e58c4c3225e2de2039)

Josmithr replied:
@jikim-msft FYI

Josmithr committed:
improvement: Don't fail when we encounter non ISharedObject data

Josmithr committed:
build: Update lock file

Josmithr committed:
style: Prettier

Josmithr committed:
improvement: Better type-wise handling of unknown objects

Josmithr committed:
docs: Fix typo

Co-authored-by: Alex Villarreal <716334+alexvy86@users.noreply.github.com>

Josmithr commented:
> Approved with minor concern: During testing of different data types, casting some data objects to ISharedObject may fail, potentially resulting in errors and causing the page to break.

Good call! And the case that was a failure mode also represents a case where we probably want to display some disclaimer to the user on the extension side. I added an `UnknownObjectNode` type to the tree grammar to represent this case.

@jikim-msft FYI, since you'll have to handle this case on the extension side.

Josmithr committed:
docs: Update terminology

Josmithr committed:
docs: Clarify terminology

Josmithr committed:
refactor: Remove unecessary string literals

Josmithr committed:
fix: De-register event listener before disposing dataVisualizer

Josmithr committed:
docs: Remove obsolete TODO

Josmithr committed:
refactor: Use inline constructor properties

Josmithr committed:
revert: 1735c07f72291ca7cfb0970500520cb4e7dd8521

Josmithr committed:
refactor: Make "nodeKind" more flexible for forwards/backwards compat

Josmithr commented on packages/tools/client-debugger/client-debugger/src/data-visualization/VisualTree.ts:
```diff
@@ -0,0 +1,233 @@
+ *
+ * @public
+ */
+export enum VisualNodeKind {
```
I went ahead and made the `nodeKind` property allow arbitrary strings in addition to the enum value. That way we have an exit hatch for enum value changes. Changes to the enum are still technically breaking at the type-system level, but they won't break the node trees. Let me know what you think of this approach; definitely happy to consider alternatives.

alexvy86 replied:
I think we're still at a point where we can potentially change this in the short term if we need. So at least for MVP purposes this seems ok. I can't really argue strongly agianst an enum, tbh, I'd need to look into the details of what we've encountered before.

Josmithr committed:
refactor: Refactor tree grammar to make "label" a parent node concept, rather than a child node concept.

Simplifies the code fairly significantly, and fixes a caching bug in the visualizer graph.

Josmithr committed:
docs: Update overview diagrams to match new type signatures

Josmithr committed:
build: Revert pnpm lock to v5.4

Josmithr committed:
revert: Unwanted pnpm-lock changes

Josmithr committed:
Merge branch 'main' into client-debugger/schema-based-data-prototype

alexvy86 commented on packages/tools/client-debugger/client-debugger/src/data-visualization/Overview.md:
```diff
@@ -2,16 +2,23 @@
+
+-   "Application": The client application, which initializes the "debugger".
+-   "Debugger": Represents the client-side debugger, which communicates application state updates via message passing to the "consumer".
+-   "Consumer": External tooling which consumes data from the "debugger" via message passing.
```
I'd add a note about the browser extension as an example of a consumer to drive it home. I know it's a separate thing and people could decide to use this piece on their own, but from our standpoint I think we can have them reference each other and it's ok.

Josmithr replied:
Yeah, that's a good call. I don't mind calling out examples, so long as it's clear that's all they are. And I think it helps lend clarity here. Will update.

alexvy86 reviewed (approved):
Minor thing, and a grammar thing still open from before, I believe? But 🚀 .

Josmithr committed:
docs: Fix typo

Co-authored-by: Alex Villarreal <716334+alexvy86@users.noreply.github.com>

Josmithr committed:
docs: Add example to "consumer" definition

Josmithr commented:
> Minor thing, and a grammar thing still open from before, I believe? But 🚀 .

Totally missed that lingering comment. I cleaned that up as well. Thanks!
